openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java

```diff
@@ -414,6 +419,193 @@ public Response search(SearchRequest request) throws IOException {
+        throw new SearchException(String.format("Search failed due to %s", e.getMessage()));
+      }
+    }
+  }
+
+  public List<GlossaryTerm> buildGlossaryTermNestedHierarchy(SearchResponse searchResponse) {
+    Map<String, GlossaryTerm> termMap = new LinkedHashMap<>(); // termMap represent glossary terms
+    Map<String, GlossaryTerm> rootTerms = new LinkedHashMap<>(); // rootTerms represent glossaries
+
+    for (var hit : searchResponse.getHits().getHits()) {
+      Map<String, Object> hitSourceMap = new HashMap<>(JsonUtils.getMap(hit.getSourceAsMap()));
+
+      GlossaryTerm term = createTermFromMap(hitSourceMap);
+      Map<String, Object> glossaryInfo = (Map<String, Object>) hitSourceMap.get("glossary");
+
+      if (glossaryInfo != null) {
+        GlossaryTerm parentTerm = createTermFromMap(glossaryInfo);
+        rootTerms.putIfAbsent(parentTerm.getFullyQualifiedName(), parentTerm);
+      } else {
+        Map<String, Object> parentInfo = (Map<String, Object>) hitSourceMap.get("parent");
+        GlossaryTerm parentTerm = createTermFromMap(parentInfo);
+        termMap.putIfAbsent(parentTerm.getFullyQualifiedName(), parentTerm);
+      }
+
+      termMap.putIfAbsent(term.getFullyQualifiedName(), term);
+    }
+
+    termMap.putAll(rootTerms);
+
+    termMap
+        .values()
+        .forEach(
+            term -> {
+              String parentFQN = getParentFQN(term.getFullyQualifiedName());
+              String termFQN = term.getFullyQualifiedName();
+
+              if (parentFQN != null && termMap.containsKey(parentFQN)) {
+                GlossaryTerm parentTerm = termMap.get(parentFQN);
+                List<EntityReference> children = parentTerm.getChildren();
+                EntityReference termRef = term.getEntityReference();
+                termRef.setChildren(term.getChildren());
+                children.add(termRef);
+                children.sort(Comparator.comparing(EntityReference::getName));
+                parentTerm.setChildren(children);
+              } else {
+                if (rootTerms.containsKey(termFQN)) {
+                  GlossaryTerm rootTerm = rootTerms.get(termFQN);
+                  rootTerm.setChildren(term.getChildren());
+                }
+              }
+            });
+
+    return new ArrayList<>(rootTerms.values());
+  }
+
+  private GlossaryTerm createTermFromMap(Map<String, Object> termInfo) {
+    GlossaryTerm term = new GlossaryTerm();
+    if (termInfo != null) {
+      term.setId(UUID.fromString(termInfo.get("id").toString()));
+      term.setName(termInfo.get("name").toString());
+      term.setDisplayName(
+          termInfo.get("displayName") != null
+              ? termInfo.get("displayName").toString()
+              : termInfo.get("name").toString());
+      term.setFullyQualifiedName(termInfo.get("fullyQualifiedName").toString());
+      term.setChildren(new ArrayList<>());
+    }
+    return term;
+  }
+
+  private String getParentFQN(String fullyQualifiedName) {
```

Comment:
we have resusable methods in FullyQualifiedName.java to get parent FQNs and handle them more efficiently , rather than doing it manually and checking on ' '